### PR TITLE
Bug 2056387: fix alibaba kubelet node name unit

### PR DESCRIPTION
**Pull request by twiest**
**- What I did**
- [x] Change usr-local-bin-alibaba-kubelet-nodename to generate a
       systemd environment file. This fixes bug 2056387 systemctl
       daemon-reload to alibaba-kubelet-nodename.service.yaml.
- [x] Add a systemd generator so that the kubelet.service will source
       the env file generated by usr-local-bin-alibaba-kubelet-nodename.
- [x] Test the new unit file fixes bug 2056387.

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2056387

**- How to verify it**
1. Ensure the file gets placed: /usr/local/bin/alibaba-kubelet-nodename
1. Ensure the file gets placed: /usr/local/lib/systemd/system-generators/alibaba-kubelet-extra-env-generator.sh
1. Reload systemd with "systemctl daemon-reload"
1. Ensure that kubelet.service now lists a generator with "systemctl status kubelet". Should look something like:
```
● kubelet.service - Kubernetes Kubelet
   Loaded: loaded (/etc/systemd/system/kubelet.service; enabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/kubelet.service.d
           └─10-mco-default-madv.conf, 20-logging.conf
           /run/systemd/generator.early/kubelet.service.d
           └─alibaba-extra-env.conf
```

Additionally, try adding a rhel node to an existing cluster on Alibaba by adapting these instructions:
https://docs.openshift.com/container-platform/4.11/machine_management/adding-rhel-compute.html

**- Description for the changelog**
Changed /usr/local/bin/alibaba-kubelet-nodename to output and env file and added a generator so that kubelet.service consumes that env file.

### CODE REVIEW

**Comment by cgwalters** [comment]
Hmm.  This file should probably be a systemd generator.  It's frowned upon to have the execution of systems units dynamically change the system state, because that creates the potential for race conditions and loops.  It's also slower because it forces systemd to reread all units.

Anyways...for now though,
/approve
of the patch as looking correct to me, but this code needs cleanup

**Comment by elmiko** [comment]
/retitle Bug 2056387: fix alibaba kubelet node name unit

File: templates/common/alibabacloud/files/usr-local-lib-systemd-system-generators-alibaba-kubelet-nodename-generator.sh.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    kubelet_service_d_dir="$earlydir/kubelet.service.d"
+    node_conf="$kubelet_service_d_dir/alibaba-node-name.conf"
+
+    # Set KUBE_NODE_NAME to be overridden by --hostname-override in the kubelet.
+    # Format required is "<region_id>.<instance_id>"
+    # https://github.com/kubernetes/cloud-provider-alibaba-cloud/blob/master/docs/getting-started.md
+    meta_ep=http://100.100.100.200/latest/meta-data
+    region_id=$(curl -s ${meta_ep}/region-id)
+    instance_id=$(curl -s ${meta_ep}/instance-id)
```
Comment (cgwalters) [inline templates/common/alibabacloud/files/usr-local-lib-systemd-system-generators-alibaba-kubelet-nodename-generator.sh.yaml:29]: Hmm actually this is tricky because systemd generators run *really* early and it's possible we won't have networking up then.  So we'd want to retry.

I think where this is going actually is to append to an environment file...in fact, we already have a pile of these:

```
  EnvironmentFile=-/etc/kubernetes/kubelet-workaround
  EnvironmentFile=-/etc/kubernetes/kubelet-env
  EnvironmentFile=/etc/node-sizing.env
```

This file will be read only when systemd goes to start kubelet - this way we can do what we need to as a unit that runs `Before=kubelet.service` as before.

To say it another way, by writing to an `EnvironmentFile` instead of mutating `Environment=` we don't need to reload systemd.

**Comment by twiest** [comment]
@cgwalters I converted it to a systemd generator. You're correct that this is a much better way to generate this variable extension unit.

Looking through the source, it seems that there aren't any references to the usr-local-bin-alibaba-kubelet-nodename.yaml and alibaba-kubelet-nodename.service.yaml template files that I removed.

Am I correct in assuming that I don't need to update a string somewhere to now reference usr-local-lib-systemd-system-generators-alibaba-kubelet-nodename-generator.sh.yaml?

In other words, is it simply enough that this file exists in this path in order to have it sent down to the node?

Thanks!

File: templates/common/alibabacloud/files/usr-local-lib-systemd-system-generators-alibaba-kubelet-nodename-generator.sh.yaml (hunk 1 above)
Comment (cgwalters) [reply templates/common/alibabacloud/files/usr-local-lib-systemd-system-generators-alibaba-kubelet-nodename-generator.sh.yaml:29]: To do this slightly more elegantly, you could have the generator synthesize the unit fragment

```
[Service]
EnvironmentFile=/run/kubelet-alibaba.env
```

Then the systemd unit would populate that file.

Notice that we don't have `-` here - we would fail if it doesn't exist.

**Comment by twiest** [comment]
@cgwalters I updated the code in this PR to hopefully reflect what you're looking for.

This code, however, doesn't fix the bug (same behavior as before this code). I have no idea what I'm doing wrong, but I cannot get systemd to use the value that's in the (immediately before written) environment file. I've tried everything I can think of to make it work.


Here is the log entry that fetches the data and writes the environment file:
```
Sep 28 00:47:47 twiest-zxgz9-rhel-worker-us-east-1b-cbd33 systemd[1]: Started Fetch the region and isntance id from Alibaba M>
```

Here is the service entry that uses the environment file when it starts up:
```
Sep 28 00:47:50 twiest-zxgz9-rhel-worker-us-east-1b-cbd33 systemd[1]: Starting Kubernetes Kubelet...
[... snip ...]
kube[1734]: I0928 00:47:50.463909    1734 flags.go:64] FLAG: --provider-id="alicloud://"
```

As you can see, the file is written 3 seconds before the kubelet starts up, but the kubelet isn't passed the environment variable as a command line variable like it should. The 2nd line above should be:
```
--provider-id="alicloud://us-east-1.i-0xi1yngjbb8mq3ngkc5m"
```

The [systemd.exec man page](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#EnvironmentFile=) says this:
```
The files listed with this directive will be read shortly before the process is executed (more
specifically, after all processes from a previous unit state terminated. This means you can
generate these files in one unit state, and read it with this option in the next. The files are
read from the file system of the service manager, before any file system changes like
bind mounts take place).
```

So you are correct that this should work, I just don't know why it's not working.

I've got some ideas that I'm going to try out tomorrow. LMK if you can tell what I'm doing wrong.

File: templates/common/alibabacloud/files/usr-local-bin-alibaba-kubelet-nodename.yaml [hunk 2]
```diff
@@ -5,10 +5,10 @@ contents:
     #!/bin/bash
     set -e -o pipefail
```
Comment (cgwalters) [inline templates/common/alibabacloud/files/usr-local-bin-alibaba-kubelet-nodename.yaml:6]: I'd use full [bash strict mode](http://redsymbol.net/articles/unofficial-bash-strict-mode/) as a general rule, specifically we want `-u` here.

File: templates/common/alibabacloud/files/usr-local-bin-alibaba-kubelet-nodename.yaml [hunk 3]
```diff
@@ -20,7 +20,6 @@ contents:
     INSTANCEID=$(curl -s ${META_EP}/instance-id)
 
    
-    cat > "${NODECONF}" <<EOF
-    [Service]
-    Environment="KUBELET_NODE_NAME=${REGIONID}.${INSTANCEID}"
+    cat > "${ENV_FILE}" <<EOF
+    KUBELET_NODE_NAME=${REGIONID}.${INSTANCEID}
```
Comment (cgwalters) [inline templates/common/alibabacloud/files/usr-local-bin-alibaba-kubelet-nodename.yaml:24]: Related to the above, is it possible that `REGIONID` is unset because something else went wrong?

File: templates/common/alibabacloud/files/usr-local-lib-systemd-system-generators-alibaba-kubelet-extra-env-generator.sh.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+mode: 0755
+path: "/usr/local/lib/systemd/system-generators/alibaba-kubelet-extra-env-generator.sh"
+contents:
+  inline: |
+    #!/bin/sh
+    set -e -o pipefail
```
Comment (cgwalters) [inline templates/common/alibabacloud/files/usr-local-lib-systemd-system-generators-alibaba-kubelet-extra-env-generator.sh.yaml:6]: Ditto here re strict mode.  

Also note https://github.com/coreos/fedora-coreos-config/pull/1797/commits/1a204418d187834453dea01d991b3c3353937b98
specifically https://github.com/coreos/fedora-coreos-config/blob/testing-devel/overlay.d/05core/usr/lib/coreos/generator-lib.sh#L3
to ensure we get log messages if something goes wrong in this code.

**Review by cgwalters — COMMENTED** [review]
Not sure what's going wrong, it looks like it should work.  Perhaps the added logging/strict mode will pop out a typo we're both missing.

**Comment by jianli-wei** [comment]
@twiest please see https://bugzilla.redhat.com/show_bug.cgi?id=2056387#c32 for the installation issues, thanks!

**Comment by twiest** [comment]
> @twiest please see https://bugzilla.redhat.com/show_bug.cgi?id=2056387#c32 for the installation issues, thanks!

Replied in the bug.

File: templates/common/alibabacloud/files/usr-local-bin-alibaba-kubelet-nodename.yaml (hunk 3 above)
Comment (twiest) [reply templates/common/alibabacloud/files/usr-local-bin-alibaba-kubelet-nodename.yaml:24]: Your suggested "strict mode" and "generator logging" changes uncovered the problem.

Here's the timeline:
1st generator run outputs this:
```
Sep 29 22:28:28 twiest-zxgz9-rhel-worker-us-east-1b-cbd33 unknown: BEGIN generator alibaba-kubelet-extra-env-generator.sh
Sep 29 22:28:28 twiest-zxgz9-rhel-worker-us-east-1b-cbd33 /usr/local/lib/systemd/system-generators/alibaba-kubelet-extra-env-generator.sh: line 26: cannot create temp file for here-document: Read-only file system
Sep 29 22:28:28 twiest-zxgz9-rhel-worker-us-east-1b-cbd33 systemd[733]: /usr/local/lib/systemd/system-generators/alibaba-kubelet-extra-env-generator.sh failed with exit status 1.
```

The generator eventually runs again, but it runs AFTER the kubelet service has started:

Kubetlet service start:
```
Sep 29 22:28:33 twiest-zxgz9-rhel-worker-us-east-1b-cbd33 systemd[1]: Starting Kubernetes Kubelet...
```

2nd generator run:
```
Sep 29 22:31:00 twiest-zxgz9-rhel-worker-us-east-1b-cbd33 unknown: BEGIN generator alibaba-kubelet-extra-env-generator.sh
Sep 29 22:31:00 twiest-zxgz9-rhel-worker-us-east-1b-cbd33 unknown: END generator alibaba-kubelet-extra-env-generator.sh
```


I'm going to work on a version of the generator that doesn't use a here-doc and re-run the test.

Thanks!

**Comment by cgwalters** [comment]
Ah so the generator was trying to create a random temporary file at the toplevel `/`.  That's a perfect example of why ostree is strict about writable state - on a traditional rpm/dpkg/etc system then you'd just end up with this cruft file that sits around mysteriously.  I've seen people do this type of bug in RPM post scripts too, ending up random strange files in `/`.

Anyways, CI here doesn't cover this so I'm good to do lgtm; did this latest code work for you and it's ready to go?

**Comment by twiest** [comment]
> Anyways, CI here doesn't cover this so I'm good to do lgtm; did this latest code work for you and it's ready to go?

Yes, my manual testing worked. I think this is ready to go.

@jianli-wei I've removed WIP from the title. I think this is ready to test.
